### PR TITLE
Add option to pre-download checkpoint to load

### PR DESCRIPTION
**Pull request by epwalsh**
This should help with the situation @dirkgr, @2015aroras, and I discussed. You can activate this with:

```
--trainer.checkpointer.pre_download=true
```

### CODE REVIEW

File: src/olmo_core/train/checkpoint.py [hunk 1]
```diff
@@ -128,6 +154,20 @@ def load(
         """
         dir = normalize_path(dir)
 
+        if is_url(dir) and self.pre_download:
+            target = self.work_dir / "load" / os.path.basename(dir)
+            if get_fs_local_rank() == 0:
+                copy_dir(dir, target, save_overwrite=self.save_overwrite)
+            barrier()
```
Comment (dirkgr) [inline src/olmo_core/train/checkpoint.py]: This is means that if the copy takes longer than 10 minutes, we will definitely timeout. Is there a way to extend the timeout, or have the other ranks participate in the copying somehow?

Comment (epwalsh) [reply src/olmo_core/train/checkpoint.py]: In your PR for the 32B config you extended the timeout to 30 minutes. Is that enough?

Comment (dirkgr) [reply src/olmo_core/train/checkpoint.py]: Yeah, but 30 minutes is really bad when you don't need it. Might make a 1B model stall out for 30 minutes when it doesn't need to.

Comment (epwalsh) [reply src/olmo_core/train/checkpoint.py]: There's a way to decrease the timeout later. Let me see if I can get that working.

Comment (epwalsh) [reply src/olmo_core/train/checkpoint.py]: actually we can just increase the timeout for the checkpointing PG: d67f767

**Comment by 2015aroras** [comment]
While this works, it seems like a short-term solution. It makes a model copy per node. This is good when doing HSDP with a replica per node (which is our current need), but bad for FSDP. Ideally we would only download the files needed by the rank or node.

**Comment by epwalsh** [comment]
> While this works, it seems like a short-term solution. It makes a model copy per node. This is good when doing HSDP with a replica per node (which is our current need), but bad for FSDP.

If I understood correctly, the original issue this is trying solve is loading a checkpoint that was saved with different sharding configuration (e.g. saved with FSDP, load with HSDP) because that leads to too many small requests to GCS. If that's the case then there's no point in having this toggled on when you're loading a checkpoint saved with the same sharding configuration.

> Ideally we would only download the files needed by the rank or node.

This is basically what happens without this option toggled on. Except due to the way `torch.distributed.checkpoint` works, we're downloaded slices of files, not whole files. In any case, each rank only downloads the data it needs to fill its local shard of parameters.

**Review by 2015aroras — APPROVED** [review]

**Pull request merged**